File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.Set (fromAscList)

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit =
  sum $ foldMap f factors'
  where
    f factor = fromAscList [0, factor .. limit -1]
    factors' = filter (/= 0) factors

File: prolog/sum-of-multiples/sum_of_multiples.pl
sum_of_multiples(_, 0, 0).
sum_of_multiples(Factors, Limit, Res) :-
    Limit2 is Limit - 1,
    sum_of_multiples(Factors, Limit2, Res2),
    (multiples(Factors, Limit2),
        Res is Res2 + Limit2;
        Res is Res2
    ), !.

multiples([], _) :- false.
multiples([Factor|T], N) :-
    (N rem Factor =:= 0,
        true;
        multiples(T, N)
    ), !.